### ISSUE TITLE
[bug] Segfault and exception not caught

### ISSUE THREAD

**Issue by simogasp** [alicevision/CCTag#65]
On the CPU version, on MacOSX, segfault happens with some input images, eg these ones (among many others from that dataset):
https://github.com/alicevision/CCTagDataset/blob/develop/bench/images/synthetic/3Crowns/xp3/images/00247.png
https://github.com/alicevision/CCTagDataset/blob/develop/bench/images/synthetic/3Crowns/xp3/images/00258.png
https://github.com/alicevision/CCTagDataset/blob/develop/bench/images/synthetic/3Crowns/xp3/images/00260.png

After some debugging it turned out that the problem arises inside `cctagMultiresDetection()` and more precisely [inside this loop here](https://github.com/alicevision/CCTag/blob/develop/src/cctag/Multiresolution.cpp#L320), in which, for each candidate marker, the ellipse fitting its `EdgePoint`s is estimated. If the list of `EdgePoint` is empty (!), the algorithm goes on, it tries somehow to estimate the ellipse with `ellipseFitting()`, getting weird intermediate matrices filled with `NaN`s and somehow falling in one of the degenerated cases of `to_ellipse()`, throwing the relevant exception. Which is caught but not dealt with in the caller, ie 
```c++
 try
      {
        // this eventually throw an exception if rescaledOuterEllipsePoints.empty()
        numerical::ellipseFitting(rescaledOuterEllipse, rescaledOuterEllipsePoints);

        .....
      }
      catch (...)
      {
        // catch exception from ellipseFitting (nothing is done)
      }
```
Ok in theory this should not raise problems, and I don't know why it crashes on Mac (actually, funny story, it crashes if launched at command line, it works just fine if it is launched from within the IDE...). Putting this aside, some remarks: 

1. Is it normal that a candidate at that point of the algorithm has no edgepoints? I see that even before `outlierRemoval()` the list of point is empty so i guess it may come from upstream.

2. I'd add a simple check to see if the list of points is empty or, better, `size() < 5` and skip the candidate. Which brings to next point.

3. all the ellipse fitting functions and their wrappers shouldn't be checking whether the number of points are >= 5 before going on with the algorithm? It would be better to throw and exception or an error at that point, rather than going on with weird and useless computations. 

Of course with (2) the problem is solved and it does not crashes anymore, (3) IMHO would improve the robustness of the function (I didn't like seeing those NaN matrices... I'm a very impressionable guy... :-P). As for (1), that's on you folks as I don't know the implementation details of the algorithm :-)

I can PR (1) and (2) if you want as i already did it.

For the record it happens calling 
```
detection -i 00247.png -n 3 -o ./out
```

@lcalvet @griwodz @zvrba

**Comment by simogasp** [comment]
ok the segfault may be a corrupted osx version (???!)
https://stackoverflow.com/questions/44712703/xcode-produces-segmentation-fault-as-opposed-to-exception

**Comment by simogasp** [comment]
@zvrba 
investigating a little bit and opening a can of worms :-) i found that ellipseFitting gives wrong results with degenerate configurations of points, eg
```c++
    std::vector<Point3f> pts;
    pts.reserve(5);
    // These are points lying on two crossing lines at the origin
    pts.emplace_back( 0.0000f, 1.0f);
    pts.emplace_back( 0.0000f, -1.0f);
    pts.emplace_back( 0.0000f, 0.0f);
    pts.emplace_back( 1.0000f, 0.0f);
    pts.emplace_back( -1.0000f, 0.0f);
    cctag::numerical::geometry::Ellipse ellipse;
    cctag::numerical::ellipseFitting(ellipse, pts);
    std::cout << "a " << ellipse.a() << std::endl;
    std::cout << "b " << ellipse.b() << std::endl;
    std::cout << "angle " << ellipse.angle() << std::endl;
    std::cout << "center " << ellipse.center() << std::endl;
    const auto& m = ellipse.matrix();
    std::cout << ellipse<< std::endl;
    std::cout << m<< std::endl;
    std::cout << m.topLeftCorner<2,2>().determinant() << std::endl;
    std::cout << m.determinant() << std::endl;
    std::cout << (m(0,0)+m(1,1))*m.determinant() << std::endl;

    for (const auto& p : pts)
    {
        std::cout << p.transpose() * m * p<< std::endl;
    }
```
gives
```
a 0.894427
b 0.894427
angle 3.14159
center [0, 0, 1]
e = [ 1.25 0 0 ; 0 1.25 -0 ; 0 -0 -1 ] 
m = 1.25    0    0
   0 1.25   -0
   0   -0   -1
 m.topLeftCorner<2,2>().determinant()  = 1.5625  // required to be > 0 to be an ellipse
m.determinant() -1.5625  // required to be != 0 to be a non-degenerated conic
(m(0,0)+m(1,1))*m.determinant() -3.90625  // required to be < 0 to be a real ellipse (as opposed to imaginary)
// verifying that p' * m * p = 0
0.25
0.25
-1
0.25
0.25
```

So it seems like that it enforces to find the "closest ellipse" fitting the points instead of finding the ellipse through the 5 points (which may fail if the points are not on an ellipse). Is that a wanted implementation?

**Comment by zvrba** [comment]
Ellipse fitting needs 5 points at least, and the algorithm assumes that. Feeding it less than 5 points w/o any checks leads to undefined behavior (returning an arbitrary result or crashing). I have no idea why does it ever proceed if there are fewer than 5 points.

As for the second issue, yes, the intention is to find the closest ellipse to the input points. Points come from a digitized image and it's very improbable that they're ever going to lie exactly on an ellipse. IIRC, the algorithm finds points to fit first by following the gradient direction (and convexity criterion), and then it finds candidate points against which to fit an ellipse. A degenerate input is therefore highly improbable (I won't say impossible :))

**Comment by simogasp** [comment]
No there are no checks that prevents going on with less than 5 points. I just added them and it should be fine. Along another check inside `fit_solver` to avoid degenerate cases (eg the same point or points lying on a line...).

ok for the second, i imagined that it was like that when i saw the results of that example. I will add a comment though to make it clear

**Comment by simogasp** [comment]
any idea why a candidate may have no edgePoints though?

**Comment by zvrba** [comment]
Nope.

**Comment by simogasp** [comment]
concerning the segfault, just to put to shame my stupidity, here is what happened (for future memory and in case someone else bump into similar problems)

The premise is that i have multiple version of LLVM, the "standard" one shipped by Apple and LLVM 3.9 installed with homebrew (mainly used to have openmp support). In order to switch from one to another, in my .profile i can activate the homebrew one by adding the bin directory to PATH, its `lib` directory to `LIBRARY_PATH` and, here comes the problem, i also add the latter to the `DYLD_LIBRARY_PATH`.

Whenever i tried to execute a code throwing an exception, even the most basic one, I got a segfault instead of catching the exception. LLDB stopped with this kind of back trace
```
(lldb) thread backtrace -c 15
* thread #1: tid = 0x6876b3, 0x0000000101eff8a4 libunwind.dylib`_Unwind_RaiseException + 94, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x0000000101eff8a4 libunwind.dylib`_Unwind_RaiseException + 94
    frame #1: 0x00007fffc4e767ad libc++abi.dylib`__cxa_throw + 104
```
The problem turned out to be that `libunwind.dylib` that is linked with `libSystem.B.dylib`, both locate `/usr/lib/` (system ones), but `libunwind.dylib` is also provided by LLVM from homebrew in  its `lib` directory: since I added it to `DYLD_LIBRARY_PATH`, that version was used instead which was evidently not compatible with the system one. Or at least that is what solved the problem for me, I removed the `lib` directory from `DYLD_LIBRARY_PATH` and everything was working again.
That also explained why launching it from within the IDE the segfault was not generated.

I'll push some of those propositions (2) and (3) for review
